File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys
sys.setrecursionlimit(10**6)

n = int(input())
ground = [list(map(int, input().split())) for _ in range(n)]
result = -1e9

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
def dfs(x, y, h):
  visited[x][y] = True
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]
    if 0<=nx<n and 0<=ny<n and not visited[nx][ny] and ground[nx][ny] >= h:
      dfs(nx, ny, h)

for height in range(1, 101):
  cnt = 0
  visited = [[False] * n for _ in range(n)]

  for x in range(n):
    for y in range(n):
      if ground[x][y] >= height and not visited[x][y]:
        dfs(x, y, height)
        cnt += 1

  result = max(result, cnt)

print(result)

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
  static int[] di = {-1, 1, 0, 0};
  static int[] dj = {0, 0, 1, -1};

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(br.readLine());
    int maxHeight = 0;
    int answer = 0;
    int[][] landArr = new int[N][N];

    for (int i = 0; i < N; i++) {
      StringTokenizer st = new StringTokenizer(br.readLine());
      for (int j = 0; j < N; j++) {
        landArr[i][j] = Integer.parseInt(st.nextToken());
        maxHeight = Math.max(maxHeight, landArr[i][j]);
      }
    }

    while (maxHeight >= 0) {
      boolean[][] visited = drowning(landArr, maxHeight, N);
      int cnt = 0;
      for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
          if (!visited[i][j]) {
            cnt++;
            dfs(i, j, visited, N);
          }
        }
      }
      answer = Math.max(answer, cnt);
      maxHeight--;
    }
      
    System.out.println(answer);
  }
  public static boolean[][] drowning(int[][] landArr, int maxHeight, int N) {
    boolean[][] isMoist = new boolean[N][N];
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) {
        if (landArr[i][j] <= maxHeight) {
          isMoist[i][j] = true;
        }
      }
    }
    return isMoist;
  }
  public static void dfs(int i, int j, boolean[][] visited, int N) {
    visited[i][j] = true;
    for (int k = 0; k < 4; k++) {
      int ni = di[k] + i;
      int nj = dj[k] + j;
      if (ni < 0 || nj < 0 || ni >= N || nj >= N || visited[ni][nj]) {
        continue;
      }
      dfs(ni, nj, visited, N);
    }
  }
}